src/Php/ComposerPhpVersionFactory.php

```diff
@@ -0,0 +1,114 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Php;
+
+use Composer\Semver\VersionParser;
+use Nette\Utils\Json;
+use Nette\Utils\JsonException;
+use Nette\Utils\Strings;
+use PHPStan\File\CouldNotReadFileException;
+use PHPStan\File\FileReader;
+use function count;
+use function end;
+use function is_file;
+use function is_string;
+use function sprintf;
+
+class ComposerPhpVersionFactory
+{
+
+	private ?PhpVersion $minVersion = null;
+
+	private ?PhpVersion $maxVersion = null;
+
+	/**
+	 * @param string[] $composerAutoloaderProjectPaths
+	 */
+	public function __construct(
+		private array $composerAutoloaderProjectPaths,
+		?int $minVersion,
+		?int $maxVersion,
+	)
+	{
+		if ($minVersion !== null) {
+			$this->minVersion = new PhpVersion($minVersion);
+		}
+		if ($maxVersion !== null) {
+			$this->maxVersion = new PhpVersion($maxVersion);
+		}
+
+		if ($minVersion !== null && $maxVersion !== null) {
+			return; // use values from config files
+		}
+
+		// fallback to composer.json based php-version constraint
+		$composerPhpVersion = $this->getComposerRequireVersion();
+		if ($composerPhpVersion === null) {
+			return;
+		}
+
+		$parser = new VersionParser();
+		$constraint = $parser->parseConstraints($composerPhpVersion);
+
+		if ($this->minVersion === null && !$constraint->getLowerBound()->isZero()) {
+			$minVersion = $this->buildVersion($constraint->getLowerBound()->getVersion());
+			// allow checks with < below lower bound to prevent errors like
+			// 'Comparison operation "<" between int<70205, 90000> and 70205 is always false.'
+			if ($minVersion !== null) {
+				$this->minVersion = new PhpVersion($minVersion->getVersionId() - 1);
```

Comment:
This subtraction seems very dodgy to me. If I get it correctly this is to fix warnings where `if (PHP_VERSION_ID < 70100)` where I have a require of php `>=7.1` for example. But why should it not warn? Because clearly per the requirement the code should not be allowed to run with a version lower than 70100.

Reply:
Re-reading your comment above:

> it make sense to check for php-versions smaller then the composer.json declared supported ones, to be super safe.

Yes it may make sense in some cases, but then you want to explicitly override PHPStan and add this in the baseline IMO. In most cases I'd think you will want PHPStan to tell you that this check can be removed when you bump your PHP version requirement, which suddenly would not work anymore here due to the -1.

Reply:
relevant examples from previous CI failures are

- https://github.com/phpstan/phpstan-src/actions/runs/8267804971/job/22619260364?pr=2968 (in composer)
- https://github.com/phpstan/phpstan-src/actions/runs/8267804971/job/22619308636?pr=2968 (in PrestaShop)

I am fine with leaving this error and let the projects put them into a baseline, but my thinking was, that it might be pretty common that code tries to detect whether it is invoked from a not supported php-version (so whether PHP_VERSION is below the lower limit of composer.json requirements)

Reply:
Right yeah those two examples for Composer are definitely things I'd be ok ignoring if it gets me visibility on other things I forgot to delete.

Reply:
Yeah, please remove the `-1`.